lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1527,6 +1527,258 @@ bool OptimizeBatchNorm::run(Function *F, const CompilationContext &cctx) {
+    for (n = 0; n < tile.getType()->dims()[(chId + 1) % 4]; n++) {
+      for (h = 0; h < tile.getType()->dims()[(chId + 2) % 4]; h++) {
+        for (w = 0; w < tile.getType()->dims()[(chId + 3) % 4]; w++) {
+          std::vector<glow::dim_t> dimsE = {c, n, h, w};
+          std::rotate(dimsE.begin(), dimsE.begin() + dimsE.size() - chId,
+                      dimsE.end());
+          if (handle.at(llvm::ArrayRef<glow::dim_t>(dims)) != expected) {
+            return false;
+          }
+        }
+      }
+    }
+    result[c] = expected;
+  }
+  return true;
+}
+
+/// Collects the longest chain of arithmetic operations with constants starting
+/// from \p start. Updates \p scale and \p bias as it collects the operands.
+/// \p returns the last node in the chain.
+static NodeValue collectArithmeticChain(Function *F, NodeValue start,
+                                        Constant &scale, Constant &bias,
+                                        int32_t chIdx) {
+
+  Node *user = getOnlyUser(*start.getNode());
+  NodeValue chainEnd = start;
+
+  auto isSupportedForMerge = [](const Node *n) {
+    return isa<MulNode>(n) || isa<AddNode>(n) || isa<SubNode>(n) ||
+           isa<DivNode>(n);
+  };
+
+  while (user && isSupportedForMerge(user)) {
+    auto lhs = user->getNthInput(ArithmeticNode::LHSIdx);
+    auto rhs = user->getNthInput(ArithmeticNode::RHSIdx);
+
+    // Paranoid.
+    assert(((lhs == chainEnd) || (rhs == chainEnd)) && "Not a user?");
+
+    auto out = user->getNthResult(ArithmeticNode::ResultIdx);
+
+    // Quantized arithmetic operations may change scale of result, we don't want
+    // to deal with that. May be supported later if needed.
+    if (lhs.getType() != out.getType() || rhs.getType() != out.getType()) {
+      break;
+    }
+
+    // Only take this one if its other argument is a constant.
+    // TODO: We can also support Splat here if needed.
+    auto *c = dyn_cast<Constant>(lhs == chainEnd ? rhs : lhs);
+    if (!c) {
+      break;
+    }
+
+    const dim_t numChannels = c->dims()[chIdx];
+
+    std::vector<float> toMerge(c->dims()[chIdx]);
+    if (!isConstBroadcasted(toMerge, *c, chIdx)) {
+      break;
+    }
+
+    auto biasH = bias.getPayloadMutable().getHandle();
+    auto scaleH = scale.getPayloadMutable().getHandle();
+
+    for (dim_t i = 0; i < numChannels; i++) {
+      if (isa<DivNode>(user)) {
+        scaleH.raw(i) /= toMerge[i];
+        biasH.raw(i) /= toMerge[i];
+      } else if (isa<MulNode>(user)) {
+        scaleH.raw(i) *= toMerge[i];
+        biasH.raw(i) *= toMerge[i];
+      } else if (isa<SubNode>(user)) {
+        // TODO: Can we support Sub(Constant, Chain)?
+        if (chainEnd == rhs) {
+          break;
+        }
+        biasH.raw(i) -= toMerge[i];
+      } else if (isa<AddNode>(user)) {
+        biasH.raw(i) += toMerge[i];
+      } else {
+        llvm_unreachable("Unsupported type!");
+      }
+    }
+    chainEnd = user->getNthResult(0);
```

Comment:
Can we assert one line above that `user->isArithmetic()` and then use `ArithmeticNode::ResultIdx` here?